File: src/java/hash/36.valid-sudoku.java
import java.util.*;

class P36 {
    // @lc app=leetcode id=36 lang=java
    // @lc code=start
    public class Solution {
    public boolean isValidSudoku(char[][] board) {

        if(board==null || board.length ==0 || board[0].length ==0)
            return false;

        List<Set<Character>> rows = new ArrayList<>();
        List<Set<Character>> cols = new ArrayList<>();

        List<List<Set<Character>>> gridLists = new ArrayList<>();

        for(int i=0; i< 9; i++){
           rows.add(new HashSet<>()) ;   
           cols.add(new HashSet<>());
        }

        for (int i =0; i<3;i++){
            List<Set<Character>>  list = new ArrayList<>();
            gridLists.add(list);
            
            for(int j =0; j <3; j++) {
                list.add(new HashSet<>()) ; 
            }

        }


        for(int i =0; i <9;i++) {
            for (int j =0; j <9; j++) {

                char elem = board[i][j];

                if(elem == '.') {
                    continue;
                }
           
                Set<Character> rowSet = rows.get(i);
                Set<Character> colsSet = cols.get(j);
                boolean newElem = rowSet.add(elem);
                if(!newElem) {
                
                    return false;
                }
                newElem = colsSet.add(elem);
                if(!newElem) {
                  
                    return false;
                }
                
                int gridRowIndex = i /3;
                int gridColIndex = j / 3;
                List<Set<Character>> list = gridLists.get(gridRowIndex);
                Set<Character> gridSet = list.get(gridColIndex);

                newElem =gridSet.add(elem);
                if(!newElem) {
                  
                    return false;
                }


            }
        }

        return true;
    }
    }
    // @lc code=end
}

File: src/go/hash/36.valid-sudoku.go
package main

/*
 * @lc app=leetcode id=36 lang=golang
 *
 * [36] Valid Sudoku
 */

// @lc code=start
func isValidSudoku(board [][]byte) bool {
	rows := make([]map[byte]bool, 9)
	cols := make([]map[byte]bool, 9)
	boxes := make([]map[byte]bool, 9)

	for i := 0; i < 9; i++ {
		rows[i] = make(map[byte]bool)
		cols[i] = make(map[byte]bool)
		boxes[i] = make(map[byte]bool)
	}

	for r := 0; r < 9; r++ {
		for c := 0; c < 9; c++ {
			if board[r][c] != '.' {
				val := board[r][c]
				boxIndex := (r/3)*3 + c/3

				if rows[r][val] || cols[c][val] || boxes[boxIndex][val] {
					return false
				}

				rows[r][val] = true
				cols[c][val] = true
				boxes[boxIndex][val] = true
			}
		}
	}

	return true
}

// @lc code=end
